bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -668,6 +681,83 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
   }
 }
 
+bool ReplicaImp::validatePreProcessedResultSignatures(preprocessor::PreProcessedResultMsg &msg) {
+  auto [buf, buf_len] = msg.getResultSignaturesBuf();
+  auto sigs = preprocessor::PreProcessResultSignature::deserializeResultSignatureList(buf, buf_len);
+
+  // f because the primary also executes the request (check PreProcessor::numOfRequiredReplies())
+  const auto expectedSignatureCount = getReplicaConfig().fVal;
+  if (sigs.size() < expectedSignatureCount) {
+    LOG_ERROR(GL,
+              "PreProcessedResult signatures validation failure - not enough signatures. Expected "
+                  << expectedSignatureCount << " got " << sigs.size());
+    return false;
+  }
+
+  for (const auto &s : sigs) {
```

Comment:
My idea about PreProcess(ed)ResultMsg is to be just a carrier for the signatures without actually validating them.

Initially I wanted to have a validate() method which does the signature checks. However this validate is called very early in the message processing code and if it fails the message is discarded. This doesn't work for us because we want to initiate a view change if signature verification fails. 

After considering pros and cons I decided that it's better to handle this in ReplicaImp because (1) it makes more sense to have a view change triggering logic there and (2) I didn't want to couple PreProcess(ed)ResultMsg with SigManager too much. I think this will bite us in the future when we decide to rework the message encoding/decoding code.

Does this make sense?

Reply:
We all the time trying to offload the ReplicaImp code from a non-essential functionally, but the functions you added break this approach. I think it's less evil to have PreProcessResultMsg be aware of SigManager than adding all this functionality to ReplicaImp. In case a VC is required, the corresponding PreProcessResultMsg function needs just to return a flag.

Reply:
Yes, message validation can be separated from ViewChange triggering. I've fixed this.